Speedup Computation for whether Aggregates containsProbe

mwachs5 opened the pull request:
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [N/A] Did you add appropriate documentation in `docs/src`?
- [ ] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

<!--
If you PR has any impact on the user API or affects backend code generation,
please describe the change in the "Release Notes" section below.
-->

#### Type of Improvement

<!-- Choose one or more from the following (delete those that do not apply): -->
- Performance improvement


#### Desired Merge Strategy

<!-- If approved, how should this PR be merged? Delete those that do not apply -->
- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes
<!--
The title of your PR will be included in the release notes in addition to any text in this section.
Please be sure to elaborate on any API changes or deprecations and any impact on backend code generation.
-->

Add a private containsProbe var to Aggregate and use it to speed up containsProbe checks.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)` and clean up the commit message.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

seldridge reviewed (approved):
LGTM

dtzSiFive reviewed (approved):
LGTM! Caching these recursive type properties is a good call, thanks!

Only doubt is if a Data is re-used or mutated to be probe-y, but not sure that's a thing?

jackkoenig reviewed (approved):
I think this is reasonable as is but there are a couple of possible improvements:

1. Specialize this on Vec to only check the sample element
2. Consider doing it live with some normal traversal. This is tricky to figure out the right place so I'm not sure it's worth untangling at the moment, but e.g. we do that for if something contains a flipped [[1]](https://github.com/chipsalliance/chisel/blob/62ca9bdca4648ccd1d05141c0fd048d9eae72a10/core/src/main/scala/chisel3/AggregateImpl.scala#L905). That spot won't quite work because we do `containsProbe` on unbound types, but maybe we can figure out a similar approach.

dtzSiFive commented:
> Only doubt is if a Data is re-used or mutated to be probe-y, but not sure that's a thing?

https://github.com/chipsalliance/chisel/blob/62ca9bdca4648ccd1d05141c0fd048d9eae72a10/core/src/main/scala/chisel3/probe/ProbeBase.scala#L35

and uses of `setProbeModifier` (which seem today to clone)?

This will increase the object size of anything that's an Aggregate to accommodate the new field -- offhand I don't know the best recommendation but IIRC @jackkoenig went around with some tricks to reduce size needed for lazy vals in common objects? Or thereabouts?

jackkoenig commented:
> This will increase the object size of anything that's an Aggregate to accommodate the new field -- offhand I don't know the best recommendation but IIRC @jackkoenig went around with some tricks to reduce size needed for lazy vals in common objects? Or thereabouts?

I'm not too worried about an extra `Boolean` (+ 1 bit in lazy val flags but there should be space there) in Aggregates. I'd care more if this were on all `Data` but it isn't so it's fine!

jackkoenig reviewed (changes requested):
>  val ret: T = if (!data.mustClone(prevId)) data else data.cloneType.asInstanceOf[T] 

This is a problem though, good eye @dtzSiFive

@mwachs5 please add a test for this, I'm struggling to suggest the precise check but given an Aggregate that is taking advantage of the !mustClone, it will have its `containsProbe` memoized to false but then it will be mutated to itself be a Probe.

I think there are a couple of ways to fix this.
1. Do the `.containsProbe` check after "maybe clone" part. But I don't think this is the best way because I think this mix of laziness and mutation could come up in a different place some day.
2. Change the value we're memoizing to be "containsChildProbe" or something since that can never change--then the `containsProbe` check is "am I a probe OR do I containChildProbe".

mwachs5 commented:
i like the "contains child probe" formulation, will do that. And will add some tests specifically for this

mwachs5 commented:
I am going with containsProbe means "is or contains a probe". Added a bunch of tests for this. I renamed the aggregate variable to be `elementsContainProbe` and optimized it for Vecs.

mwachs5 commented on core/src/main/scala/chisel3/AggregateImpl.scala:
```diff
@@ -233,6 +236,9 @@ private[chisel3] abstract class VecImpl[T <: Data] private[chisel3] (gen: => T,
 
   override def containsAFlipped = sample_element.containsAFlipped
 
+    /** Save this result, as it can be very expensive to determine this otherwise */
```
just realized my ` ./mill __.reformat` command failed, looking into it

jackkoenig reviewed (approved):
Looks great, thanks!

Changed milestone to 6.x because probes didn't exist until then.
